### ISSUE TITLE
File::Temp: umask handling

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8745]
Migrated from [rt.perl.org#41295](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=41295) (status was 'resolved')

Searchable as RT41295$

**Comment by p5pRT** [comment]
### From Peter.Dintelmann@dresdner-bank.com

This is a bug report for perl from Peter\.Dintelmann@&#8203;dresdner\-bank\.com\,
generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.

During the creation of temporary files and directories
with File&#8203;::Temp the umask is changed\. This behaviour
is not thread\-safe\.

1\. temp files
Filehandles for temp files are created r/w and directory
entries are created with mode 0600\. In the rare cases where
the filename is needed and the current umask intersects with
0600 the caller may call chmod\(\) himself\.
Thus changing the umask is not necessary\.

2\. temp directories
A newly created temp directory needs mode 0700 in order
to create a filetree underneath it\. This can be accomplished
with chmod\(\) \(thanks to Gisle Aas\)\.

3\. documentation
Contrary to the documentation leaving the umask "as is"
is not a security\-issue\, because the files are created
with mode 0600\.

The following patch tries to achieve this&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- /home/peter/blead/lib/perl5/5.9.5/File/Temp.pm      2007-01-14 11:41:34.000000000 +0100
+++ /var/tmp/Temp.pm    2007-01-15 11:08:16.000000000 +0100
@@ -483,12 +483,6 @@
       #      but may have O_NOINHERIT. This may or may not be in Fcntl.
       local $^F = 2;

-      # Store callers umask
-      my $umask = umask();
-
-      # Set a known umask
-      umask(066);
-
       # Attempt to open the file
       my $open_success = undef;
       if ( $^O eq 'VMS' and $options{"unlink_on_close"} && !$KEEP_ALL) {
@@ -502,17 +496,10 @@
        $open_success = sysopen($fh, $path, $flags, 0600);
       }
       if ( $open_success ) {
-
-       # Reset umask
-       umask($umask) if defined $umask;
-
        # Opened successfully - return file handle and name
        return ($fh, $path);

       } else {
-       # Reset umask
-       umask($umask) if defined $umask;
-
        # Error opening file - abort with error
        # if the reason was anything but EEXIST
        unless ($!{EEXIST}) {
@@ -525,24 +512,14 @@
       }
     } elsif ($options{"mkdir"}) {

-      # Store callers umask
-      my $umask = umask();
-
-      # Set a known umask
-      umask(066);
-
       # Open the temp directory
       if (mkdir( $path, 0700)) {
        # created okay
-       # Reset umask
-       umask($umask) if defined $umask;
+       # in case of odd umask
+       chmod(0700, $path);

        return undef, $path;
       } else {
-
-       # Reset umask
-       umask($umask) if defined $umask;
-
        # Abort with error if the reason for failure was anything
        # except EEXIST
        unless ($!{EEXIST}) {
@@ -2066,10 +2043,9 @@
 =item STANDARD

 Do the basic security measures to ensure the directory exists and
-is writable, that the umask() is fixed before opening of the file,
-that temporary files are opened only if they do not already exist, and
-that possible race conditions are avoided.  Finally the L<unlink0|"unlink0">
-function is used to remove files safely.
+is writable, that temporary files are opened only if they do not already
+exist, and that possible race conditions are avoided.  Finally the
+L<unlink0|"unlink0"> function is used to remove files safely.

 =item MEDIUM

---

```
</details>
Flags:
&nbsp;    category=library
&nbsp;    severity=low


---
Site configuration information for perl v5\.8\.8&#8203;:

Configured by p013176 at Mon Jul 24 10&#8203;:05&#8203;:57 MEST 2006\.

Summary of my perl5 \(revision 5 version 8 subversion 8\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=solaris\, osvers=2\.9\, archname=sun4\-solaris\-thread\-multi\-64int
&nbsp;    uname='sunos oikbm03 5\.9 generic\_118558\-28 sun4u sparc sunw\,ultra\-5\_10 '
&nbsp;    config\_args='\-Dprefix=/opt/perl32 \-Uinstallusrbinperl \-Dcc=gcc \-Doptimize=\-O2 \-Aldflags=\-R/usr/local/lib \-Duseshrplib \-Duseperlio \-Dusethreads \-Duselargefiles \-Duse64bitint \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=define use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc'\, ccflags ='\-D\_REENTRANT \-fno\-strict\-aliasing \-pipe \-Wdeclaration\-after\-statement \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-D\_REENTRANT \-fno\-strict\-aliasing \-pipe \-Wdeclaration\-after\-statement \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='3\.4\.2'\, gccosandvers='solaris2\.9'
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=87654321
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='gcc'\, ldflags =' \-R/usr/local/lib \-L/usr/local/lib '
&nbsp;    libpth=/usr/local/lib /usr/lib /usr/ccs/lib
&nbsp;    libs=\-lsocket \-lnsl \-lgdbm \-ldl \-lm \-lpthread \-lc
&nbsp;    perllibs=\-lsocket \-lnsl \-ldl \-lm \-lpthread \-lc
&nbsp;    libc=/lib/libc\.so\, so=so\, useshrplib=true\, libperl=libperl\.so
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='  \-R /opt/perl32/lib/5\.8\.8/sun4\-solaris\-thread\-multi\-64int/CORE'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-G \-R/usr/local/lib \-L/usr/local/lib'

Locally applied patches&#8203;:




---
@&#8203;INC for perl v5\.8\.8&#8203;:
&nbsp;    /opt/perl32/lib/5\.8\.8/sun4\-solaris\-thread\-multi\-64int
&nbsp;    /opt/perl32/lib/5\.8\.8
&nbsp;    /opt/perl32/lib/site\_perl/5\.8\.8/sun4\-solaris\-thread\-multi\-64int
&nbsp;    /opt/perl32/lib/site\_perl/5\.8\.8
&nbsp;    /opt/perl32/lib/site\_perl
&nbsp;    \.



---
Environment for perl v5\.8\.8&#8203;:
&nbsp;    HOME=/export/home/p013176
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/opt/perl32/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/usr/bin/bash

**Comment by p5pRT** [comment]
### From @rgs

This is solved in File&#8203;::Temp 0\.18\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'
